GCC: Add a build profile extension with the link-time optimizer enabled

fkjagodzinski opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->
### Description (*required*)

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
##### Summary of change (*What the change is for and why*)
Added an `lto` build profile extension for `GCC_ARM` toolchain.

* added `-flto` to `common` flags,
* **updated `tools/toolchains/gcc.py` to use `common` flags at link time**,
* ~marked `main()` with `MBED_USED` attribute,~
* added `-u main` to `ld` flags.

###### Results
`arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 9-2019-q4-major) 9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599]`

#### mbed-cloud-client-example@4.3.0, mbed-os-5.15.1, GCC9, Linux host
profile|RAM (data + bss)|Flash (text + data)|build time|rebuild time
---|---:|---:|---:|---:
`release`|60272|392449|141.33|10.86
`release` + `lto`|59432|356724|166.65|39.12

#### mbed-cloud-client-example@4.3.0, mbed-os-5.15.1, GCC9, Windows host
profile|RAM (data + bss)|Flash (text + data)|build time|rebuild time
---|---:|---:|---:|---:
`release`|60272|392449|1071.52|24.60
`release` + `lto`|59432|356724|1396.12|155.88

###### Build commands used to produce the above results
```bash
cd $(mktemp -d)
git clone git@github.com:ARMmbed/mbed-cloud-client-example.git
cd mbed-cloud-client-example/
git checkout 4.3.0 
git clean -dxff && git reset --hard
cp ../mbed_cloud_dev_credentials.c .

# LTO branch on top of mbed-os-5.15.0
echo 'https://github.com/fkjagodzinski/mbed-os/#a3dac1943f3770eb287c244da0a208cb4cf9b559' > mbed-os.lib

mbed deploy
cd mbed-os/ && git log --oneline -1 && cd ..
# Define Mbed OS profile
# 1. release
\time -f%e mbed compile -t GCC_ARM -m K64F --profile mbed-os/tools/profiles/release.json
# 2. release + lto
\time -f%e mbed compile -t GCC_ARM -m K64F --profile mbed-os/tools/profiles/release.json --profile mbed-os/tools/profiles/extensions/lto.json

# More details
mbed compile -t GCC_ARM -m K64F --profile mbed-os/tools/profiles/release.json --stats-depth=100
```

##### Documentation (*Details of any document updates required*)

----------------------------------------------------------------------------------------------------------------
### Pull request type (*required*)

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [X] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results (*required*)

<!--
    Required
    For example, add test results for new target
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
----------------------------------------------------------------------------------------------------------------
### Reviewers (*optional*)

<!--
    Optional
    Request additional reviewers with @username
-->
@jamesbeyond @kjbracey-arm @maciejbocianski 

----------------------------------------------------------------------------------------------------------------
### Release Notes (*required for feature/major PRs*)

<!--
    All 3 sections are compulsory for Major PR types. For Feature PRs only the summary section is required.
    This section is automatically added to release notes. Please fill in each sub-section with sufficient detail for a user.
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types). 
-->

##### Summary of changes

Add a build profile extension with the link-time optimizer enabled for GCC_ARM toolchain.

##### Impact of changes

##### Migration actions required
* The minimal required version of the `GCC_ARM` is now the *GNU Arm Embedded Toolchain Version 9-2019-q4-major*. Earlier `GCC_ARM`versions can cause various issues when the `-flto` flag is used, e.g. a platform specific error during the final link stage on Windows hosts with GCC8.

* The `noinline` attribute has to be used for every function that must be placed into a specific section (specified with a `section(".section_name")` attribute). In general, when a function is considered for inlining, the `section` attribute is always ignored. However, with the link-time optimizer enabled, the chances for inlining are much higher because the inliner works across multiple translation units. As a result, the output sections' sizes change compared to a non-lto build. This may lead to a `section ".section_name" will not fit in region "region_name"` type errors.

* The `common` flags defined for the `GCC_ARM` toolchain are now appended to the `ld` flags during a mbed-cli build. Previously, the `common` flags were appended only to `asm`, `c` and `cxx` flags. Having the same flags for the compiler and the linker is required when using the link-time optimizer (which is the case for the `lto` build profile extension). Any options unrecognized by the linker should be moved from `common` to `asm`, `c` or `cxx`.

fkjagodzinski committed:
GCC: Enable link-time optimizer for release profile

fkjagodzinski committed:
GCC: Use 'common' flags at link time

According to GCC man:
To use the link-time optimizer, -flto and optimization options should be
specified at compile time and during the final link. It is recommended
that you compile all the files participating in the same link with the
same options and also specify those options at link time.

Additionally, move the '-g3' flag out of 'common' flags in the debug
profile. Although the '-g' is correctly ignored by the linker, the
'-glevel' is not and causes a build error "ld: unrecognized option
'-g3'".

kjbracey commented on mbed.h:
```diff
@@ -113,4 +113,6 @@ using namespace mbed;
 using namespace std;
 #endif
 
+MBED_USED int main(void);
```
If this is specifically triggered by the wrap, could you perhaps put it next to the `__wrap_main` definitions, just so we know what it's for. Doesn't feel like a big public obvious header file thing.

(There are two `__wrap_main` definitions, for RTOS and non-RTOS).

fkjagodzinski replied:
Sure. I was not sure about the correct place for this line, but I can see I forgot to ask.

fkjagodzinski replied:
Ok, I tried with [mbed_boot_gcc_arm.c](https://github.com/ARMmbed/mbed-os/blob/249d35c873ff069a2eeccf0897df3ffae1997a74/rtos/source/TARGET_CORTEX/TOOLCHAIN_GCC_ARM/mbed_boot_gcc_arm.c#L73-L78) with no luck. I think the attribute has to be added in the same translation unit as the `main()` itself. How about [mbed_toolchain.h](https://github.com/ARMmbed/mbed-os/blob/249d35c873ff069a2eeccf0897df3ffae1997a74/platform/mbed_toolchain.h#L516), would that be acceptable?

fkjagodzinski replied:
@kjbracey-arm, what do you think; is mbed_toolchain.h any better than mbed.h?

kjbracey replied:
Oh, sorry, missed this. (Then saw it, then forgot it).

I was more thinking it should be out of the way, and not visible to the app, because I'm nervous that it might cause build errors. A compiler can be a bit lax about accepting different `main` prototypes, and this locks us down to this (albeit correct) one, so it might make `void main(void)` become refused.

But if we have to have it visible at the definition of main, can't be avoided, I guess. mbed_toolchain.h is a bit better.

I'd still like to see some sort of alternative - eg can we stick a MBED_USED on wrap_main?

fkjagodzinski replied:
> I'd still like to see some sort of alternative - eg can we stick a MBED_USED on wrap_main?

I've tried that, but couldn't make it work. Either a modified main.cpp or any header it includes -- only these fix the issue for me. :disappointed:

fkjagodzinski replied:
Actually, there is a better solution -- adding the `"-u main"` to `"ld"` flags.

mark-edgeworth reviewed (commented):
The notes on this change suggest that the release profile has been changed, but the code changes also affect the debug and develop profiles too.
The move of the '-c' option from common  to the asm, c and cxx tools appears to be an unrelated change, as is the corresponding change in gcc.py.

bulislaw reviewed (dismissed):
I don't think it's a breaking change, but it can break some users. We should highlight it in release notes and maybe a blog post. Any strong feelings about develop? I think debuggability of lto builds is vastly improved in 8/9 @kjbracey-arm @pan-

fkjagodzinski committed:
GCC: Enable link-time optimizer for develop profile

jamesbeyond reviewed (dismissed):
Looks good to me

0xc0170 commented on platform/mbed_toolchain.h:
```diff
@@ -514,6 +514,13 @@ typedef int FILEHANDLE;
+ * This is the case for combined "-Wl,--wrap,main" and "-flto" flags with
+ * the GCC_ARM toolchain.
+ */
+MBED_USED int main(void);
```
This is causing errors in the build , please review latest test logs

kjbracey replied:
PS - have you checked that this is still needed with GCC 9? You added it during GCC 8 testing.

fkjagodzinski replied:
Yes, GCC9 still needs a `used` attribute for wrapped main(). I'd like to move this to compiler/linker flags but so far I've only found `-fkeep-static-functions` which wastes a lot of lto gains (58408 RAM & 384236 Flash for the release profile).

fkjagodzinski replied:
Removing `"-Wl,--wrap,main"` also solves the issue. But I suppose I'm not allowed to touch that since the bare metal code uses it. https://github.com/ARMmbed/mbed-os/blob/7fbe0c6b416033a4a89ef07fe9a63990de83242c/rtos/source/TARGET_CORTEX/TOOLCHAIN_GCC_ARM/mbed_boot_gcc_arm.c#L73-L78
https://github.com/ARMmbed/mbed-os/blob/7fbe0c6b416033a4a89ef07fe9a63990de83242c/platform/source/mbed_sdk_boot.c#L109-L113

0xc0170 replied:
we would loose `mbed_main` , wouldn't we ?
 
is wrap actually working fine with lto (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=88643) in GCC ARM 9?

fkjagodzinski committed:
GCC: Add the -u flag to keep the main() symbol

This fixes the undefined reference to 'main' that arose after adding
the "-flto" flag to compilation.

This was the case for combined "-Wl,--wrap,main" and "-flto" flags.

VeijoPesonen reviewed (approved):
Looks good to me

mark-edgeworth commented on tools/toolchains/mbed_toolchain.py:
```diff
@@ -129,6 +129,22 @@ class mbedToolchain(with_metaclass(ABCMeta, object)):
+    ASM_OBJ_INFO_LIST_NAME = 'asm_objects'
+    # The name of a JSON file used to store a list of object files compiled from ASM sources.
+    ASM_OBJ_INFO_FILE_NAME = 'asm_obj.json'
+
```
This change seems only to affect the gcc compiler. It therefore should only appear in the gcc.py module, otherwise there is a risk that an inadvertent change here would adversely affect other toolchain implementations.

fkjagodzinski replied:
I moved all the code to `gcc.py`. @mark-edgeworth, could you review this again, please?

mark-edgeworth commented on tools/toolchains/mbed_toolchain.py:
```diff
@@ -466,6 +482,32 @@ def _compile_sources(self, resources, inc_dirs=None, exclude_dirs=None):
+            resources.add_file_ref(FileType.JSON, asm_obj_info_path, asm_obj_info_path)
+        except IOError:
+            pass
+
```
Move to gcc.py

mark-edgeworth commented on tools/toolchains/mbed_toolchain.py:
```diff
@@ -744,6 +786,33 @@ def link_program(self, r, tmp_path, name):
+            except (IOError, ValueError, KeyError):
+                pass
+        reorg_objects = [o for o in objects if o in asm_objects] + [o for o in objects if o not in asm_objects]
+
```
Move into gcc.py

fkjagodzinski committed:
GCC: Move link-time optimizer flags to extensions dir

Disable the lto for the default develop and release prifiles and move
the flags to tools/profiles/extensions/lto.json profile.

Usage:
mbed compile --profile release --profile tools/profiles/extensions/lto.json

jamesbeyond reviewed (dismissed):
LGTM, I believe this needs another round of CI. @0xc0170 . Also becasue this PR is not going to enabled LTO by default, so This is not a `BREAKING CHANGE` anymore

mark-edgeworth commented on tools/profiles/debug.json:
```diff
@@ -1,15 +1,15 @@
                    "-Wno-unused-parameter", "-Wno-missing-field-initializers",
                    "-fmessage-length=0", "-fno-exceptions",
                    "-ffunction-sections", "-fdata-sections", "-funsigned-char",
                    "-MMD",
-                   "-fomit-frame-pointer", "-Og", "-g3", "-DMBED_DEBUG",
+                   "-fomit-frame-pointer", "-Og", "-DMBED_DEBUG",
                    "-DMBED_TRAP_ERRORS_ENABLED=1"],
-        "asm": ["-x", "assembler-with-cpp"],
-        "c": ["-std=gnu11"],
-        "cxx": ["-std=gnu++14", "-fno-rtti", "-Wvla"],
+        "asm": ["-c", "-g3", "-x", "assembler-with-cpp"],
+        "c": ["-c", "-g3", "-std=gnu11"],
+        "cxx": ["-c", "-g3", "-std=gnu++14", "-fno-rtti", "-Wvla"],
```
These changes do not appear to be part of the work to add a profile extension. However, they look ok.

fkjagodzinski replied:
Thanks @mark-edgeworth; I left [this commit](https://github.com/ARMmbed/mbed-os/pull/11856/commits/fb40672e4de083d04a5cea49919387a411622099) to handle the `-flto` extension, where *it is recommended that you compile all the files participating in the same link with the same options and also specify those options at link time*.

mark-edgeworth commented on tools/profiles/develop.json:
```diff
@@ -1,14 +1,14 @@
+        "common": ["-Wall", "-Wextra",
                    "-Wno-unused-parameter", "-Wno-missing-field-initializers",
                    "-fmessage-length=0", "-fno-exceptions",
                    "-ffunction-sections", "-fdata-sections", "-funsigned-char",
                    "-MMD",
                    "-fomit-frame-pointer", "-Os", "-g", "-DMBED_TRAP_ERRORS_ENABLED=1"],
-        "asm": ["-x", "assembler-with-cpp"],
-        "c": ["-std=gnu11"],
-        "cxx": ["-std=gnu++14", "-fno-rtti", "-Wvla"],
+        "asm": ["-c", "-x", "assembler-with-cpp"],
+        "c": ["-c", "-std=gnu11"],
+        "cxx": ["-c", "-std=gnu++14", "-fno-rtti", "-Wvla"],
```
again, not part of the extension work, but looks ok.

mark-edgeworth reviewed (commented):
I'm quite concerned about the implementation here. The code writes some partial compilation products during compilation to a json file somewhere, then during the link stage it reads the file and acts on the content. My concern is that the file is effective a state cache; under what circumstances is it cleared, what happens when it gets out of date? How are conflicts managed? Can this information be generated on the fly, transiently, rather than storing it? It looks like it only affects the GCC_ARM toolchain so should not be a problem for the online build service code.

mark-edgeworth commented on targets/targets.json:
```diff
@@ -1222,7 +1222,6 @@
             "CAN",
             "DEBUG_AWARENESS",
             "EMAC",
-            "ETHERNET",
```
Why is this part of this change? Sorry, didn't see this on the previous review.

fkjagodzinski replied:
Please see https://github.com/ARMmbed/mbed-os/pull/11856/commits/9fd90409410131c67955f17e6c212bf54edc84b3 and https://github.com/ARMmbed/mbed-os/pull/11856#issuecomment-559731755

fkjagodzinski replied:
Moved to #12408

mark-edgeworth reviewed (dismissed):
Request that this is submitted as a separate PR as it is unrelated to the LTO issue.

fkjagodzinski committed:
Tools: Add a workaround for the GCC_ARM & LTO bug

This is a workaround for the GCC not using the strong symbols from
C files to override the weak symbols from ASM files. This GCC bug is only
present when building with the link-time optimizer (LTO) enabled. For
more details please see:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83967

This can be fixed by changing the order of object files in the linker
command; objects providing the weak symbols and compiled from assembly
must be listed before the objects providing the strong symbols.
To keep things simple, ALL object files from ASM are listed before
other object files.

mark-edgeworth reviewed (approved):
Yes, this is a great improvement over the cache-file approach. Looks fine to me now.
